### ISSUE TITLE
`Infinite loop` in `rubocop -a` for `ruby/csv`

### ISSUE THREAD

**Issue by ShockwaveNN** [bbatsov/rubocop#5586]
Failure to perform RuboCop autocorrect in https://github.com/ruby/csv.git project

--------

## Actual behavior
Take a pretty long time
```
Infinite loop detected in /home/user/sources/csv/test/lib/test/unit.rb.
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:234:in `block in iterate_until_no_changes'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:230:in `loop'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:230:in `iterate_until_no_changes'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:201:in `do_inspection_loop'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:111:in `block in file_offenses'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:121:in `file_offense_cache'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:109:in `file_offenses'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:100:in `process_file'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:78:in `block in each_inspected_file'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:75:in `each'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:75:in `reduce'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:75:in `each_inspected_file'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:67:in `inspect_files'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/runner.rb:39:in `run'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/cli.rb:150:in `execute_runner'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/cli.rb:78:in `execute_runners'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/lib/rubocop/cli.rb:38:in `run'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/bin/rubocop:13:in `block in <top (required)>'
/home/user/.rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
/home/user/.rvm/gems/ruby-2.5.0/gems/rubocop-0.52.1/bin/rubocop:12:in `<top (required)>'
/home/user/.rvm/gems/ruby-2.5.0/bin/rubocop:23:in `load'
/home/user/.rvm/gems/ruby-2.5.0/bin/rubocop:23:in `<main>'
/home/user/.rvm/gems/ruby-2.5.0/bin/ruby_executable_hooks:15:in `eval'
/home/user/.rvm/gems/ruby-2.5.0/bin/ruby_executable_hooks:15:in `<main>'
```

## Steps to reproduce the problem

1. `git clone https://github.com/ruby/csv.git`
2. `cd csv && git checkout 2cc4f858a2dcad54dc8c5e346ae691ac9a88f2a3`
3. `rubocop -a`

## RuboCop version
```
$ rubocop -V
0.52.1 (using Parser 2.5.0.0, running on ruby 2.5.0 x86_64-linux)
```

**Comment by pocke** [comment]
Probably it is related to https://github.com/bbatsov/rubocop/issues/5547 
I can reproduce this problem with `--only BlockEndNewline` option, and the related issue cause is the `BlockEndNewline` cop.

**Comment by ericmathison** [comment]
I was able to make a reduced test case using `test/lib/test/unit.rb` (from the above csv repo).

```ruby
begin
  gem 'minitest', '< 5.0.0' if defined? Gem
rescue Gem::LoadError
end
require 'minitest/unit'
require 'test/unit/assertions'

def foo
  report.each {|r| 0 + \
  0 }
end
```

For whatever reason, this reduced case will only give the infinite loop error when explicitly specifying the `--only BlockEndNewline` option.

**Comment by bquorning** [comment]
I believe this issue has been fixed by #5590.

**Comment by ShockwaveNN** [comment]
Found another one infinite loop in `ruby/csv` - probably not related to current one, in another file https://github.com/bbatsov/rubocop/issues/5628
